### ISSUE TITLE
NNCLR calculating loss only on current gpu when using multiple gpus

### ISSUE THREAD

**Issue by zhifanwu** [vturrisi/solo-learn#229]
According to Table 7(c) of NNCLR paper, the large batch size can benefit NNCLR loss. However, in the solo-learn version of NNCLR loss, it only uses data on the current GPU when using multiple GPUs, which could degrade the performance on ImageNet-100 dataset.
I think adding `p = gather(p)` would help in solving this issue.

    nn = F.normalize(nn, dim=-1)
    p = F.normalize(p, dim=-1)
    p = gather(p)

    logits = nn @ p.T / temperature

    rank = get_rank()
    n = nn.size(0)
    labels = torch.arange(n*rank, n*(rank+1), device=p.device)

    loss = F.cross_entropy(logits, labels)

**Comment by vturrisi** [comment]
Hey @zhifanwu thanks for pointing it out. In the course of this week, I'll try these changes (and some other related to distributed that we have in mind) and see how they affect the results.

**Comment by zhifanwu** [comment]
Does the `dequeue_and_enqueue` also need `gather(z)` and `gather(target)`? If each process maintain its own queue, it seems that each GPU can only visit a subset of neighbors.

**Comment by zhifanwu** [comment]
It seems that the code above will concur some problems during training, because when I ran it on 16GPUs, the model seems divergent. The validation acc won't increase (0.1~0.3 during the first 10 epochs) and the loss will increase sometimes.
I found that:
1. It seems not to be caused by learning rate. I set learning rate = base_learning_rate * GPU * batch_size / 256. I tried base_learning_rate {0.05, 0.075, 0.1, 0.3} with batch_size = 256 and GPU = 16. All of experiments couldn't converge.
2. It seems to be caused by the large batch size. I tried different numbers of GPUs with different total batch size. The batch size of a single GPU is set to 256. When I tried {2, 4} GPUs the model can be trained in a 'right way', i.e., the loss is decreasing and the validation acc is increasing. But when I tried 8 GPUs the model began to divergent, i.e., the validation acc is not increasing, with the value around 0.1~3 for 10 epochs.
3. When I used the other `z` instead of `nn` (like SimCLR), the model worked fine. So it doesn't like there is some wrong with `get_rank` or `label`.
4. There was no `nan` value in the losses.
5. When I use the following code, the performance of the model seems normal:
```
    nn = F.normalize(nn, dim=-1)
    nn = gather(nn)
    p = F.normalize(p, dim=-1)
    p = gather(p)
    logits = nn @ p.T / temperature

    n = nn.size(0)
    labels = torch.arange(n, device=p.device)

    loss = F.cross_entropy(logits, labels)
```
The difference is the `nn` is gathered together with p. I'm not sure whether this implementation is also a correct way.

Now I get stuck. It is hard for me to analyze what potential bugs cause this problem. Have you encountered this problem when implementing multi-GPU support? Really appreciate your help.

**Comment by vturrisi** [comment]
Hey @zhifanwu thanks for the update. If you gathered only for p, were you encountering problems? I think this would be in line with SimCLR implementation.

**Comment by vturrisi** [comment]
@zhifanwu I also found that the old GatherLayer had some problems and I updated it to match VISSL. Can you pull from 'gather' and check if that implementation of NNCLR works with 16 GPUs? It follows your first suggestion of gathering only `p`.

**Comment by zhifanwu** [comment]
> @zhifanwu I also found that the old GatherLayer had some problems and I updated it to match VISSL. Can you pull from 'gather' and check if that implementation of NNCLR works with 16 GPUs? It follows your first suggestion of gathering only `p`.

I've updated the `GatherLayer` and tried to train the model for 10 epochs (gathering only `p`), now it seems that the model works fine.
Thank you for your help ❤️!

I have another question: the DDP module syncs and averages gradients across all GPUs, and each GPU only calculates loss on the mini-batch of positive samples (`nn`). So does the loss need to multiply the number of GPUs to ensure the final value is on the right scale?

**Comment by vturrisi** [comment]
> I've updated the GatherLayer and tried to train the model for 10 epochs (gathering only p), now it seems that the model works fine.

Nice!

> I have another question: the DDP module syncs and averages gradients across all GPUs, and each GPU only calculates loss on the mini-batch of positive samples (nn). So does the loss need to multiply the number of GPUs to ensure the final value is on the right scale?

I'm not sure I understood what you are asking. Do you mean scaling the learning rate?

**Comment by zhifanwu** [comment]
> > I have another question: the DDP module syncs and averages gradients across all GPUs, and each GPU only calculates loss on the mini-batch of positive samples (nn). So does the loss need to multiply the number of GPUs to ensure the final value is on the right scale?
> 
> I'm not sure I understood what you are asking. Do you mean scaling the learning rate?

I mean do we need to calculate loss as:
`loss = GPU_nums * loss`
to ensure the loss is on the right scale because DDP will average gradients across all GPUs instead of adding them up.

**Comment by zhifanwu** [comment]
> It is unclear what you mean exactly. Scaling the loss by multiplying it by the number of GPUs does not make too much sense IMHO because you are already averaging over the number of samples, so the fact that you average on each GPU first and then average the results does not make any difference to the final result.
> 
> If you mean the fact that by having more GPUs you increase the batch size and you should scale the lr accordingly then yes, you are right

I think I make some mistakes, thanks for correcting me.

**Comment by vturrisi** [comment]
I'm closing since everything seems to be resolved. Feel free to re-open and thanks @zhifanwu.
